### DISCUSSION TITLE
How to Restart with the Tensor Mechanics Master Action System

### DISCUSSION THREAD

**Discussion by DaijunHU** [idaholab/moose#28944] [Q&A Modules: Solid mechanics] [answered]
Hello!

I am currently working on a two-step solid mechanics simulation, where the stress distribution obtained from the first simulation is mapped to a second model for further computation (both models share identical meshes). I am encountering two issues and would appreciate your assistance:

1. In the first simulation, I used `[Modules/TensorMechanics/Master/all]` as shown below. However, I am not sure how to carry over the stress components into the second simulation using something similar to` initial_from_file_var`. 
Is it necessary to record all stress components in AuxVariables and AuxKernels in the first simulation to achieve this restarting? 
Sorry that I couldn’t find clear parameters in the documentation for the _Tensor Mechanics Master Action System_ or _Restarting and recovering from previous simulations_.
I wanted to try with `initial_from_file_var` and `initial_from_file_timestep` in this block, but the error below actually becomes an obstacle.

2. While running the 2nd simulation using the  the following .i file, I encountered an error that left me puzzled:

> OnlyMicro80nmRestart.i:34: missing required parameter 'Modules/TensorMechanics/Master/displacements'
> Doc String: "The nonlinear displacement variables for the problem."

I actually specified `displacements = 'ux uy uz'` as shown by the attached part of my .i file, but the error persists. (I should make it clear that I did not set the displacements in my 1st simulation, so when the error occurs I added this line, but still the same error.)

Thank you for your help in advance and look forward to your reply!

Best regards,
Daijun

```
> [Mesh]
>     file = OnlyMicro80nm_out.e
>     distribution = serial
> []
> 
> [Problem]
>   restart_file_base = OnlyMicro80nm_outcp/LATEST
>   force_restart = false
> []
> 
> [Variables]
>   [./ux]
>     order = FIRST
>     family = LAGRANGE
>     initial_from_file_var = ux
>     initial_from_file_timestep = 635
>   [../]
> 
>   [./uy]
>     order = FIRST
>     family = LAGRANGE
>     initial_from_file_var = uy
>     initial_from_file_timestep = 635
>   [../]
> 
>   [./uz]
>     order = FIRST
>     family = LAGRANGE
>     initial_from_file_var = uz
>     initial_from_file_timestep = 635
>   [../]
> []
> 
> [Modules/TensorMechanics/Master/all]
>   strain = FINITE
>   displacements = 'ux uy uz'
>   add_variables = true
>   generate_output = 'stress_xx stress_yy stress_zz stress_xy stress_xz stress_yz strain_xx strain_yy strain_zz strain_xy strain_xz strain_yz'
> []
> 
> [AuxVariables]
>   [./slip_increment_1]
>    order = FIRST
>    family = MONOMIAL
>     initial_from_file_var = slip_increment_1
>     initial_from_file_timestep = 635
>   [../]
> []
```

**Comment by GiudGiud** [comment, marked as answer]
Hello

The stresses are material properties which are not restarted using ICs or using initial_from_file_var parameters. 
Your best bet is to use Checkpoint restart
https://mooseframework.inl.gov/application_usage/restart_recover.html

This second error will likely go away if you switched your syntax to the Physics syntax.
If it does not, it's asking for diplacements to be nested under Master/ instead of under Master/all

Guillaume

**Reply by DaijunHU** [reply]
Hi Guillaume,

Thank you for your reply!
I will try the Physics syntax for the 2nd error.

I am still a bit confused about the _Tensor Mechanics Master Action_. On the webpage https://mooseframework.inl.gov/syntax/Modules/TensorMechanics/Master/index.html, it is mentioned that it generates aux variables and auxkernels to aid in the output of tensor components. So I think those stress components I added in `generate_output = 'stress_xx strain_xx'` are not material properties, but a kind of _AuxVariables_ and _AuxKernels_ calculated based on the solid mechanics equations, which should follow the restarting methods of _AuxVariables_. If I use Checkpoint restart, will all these outputs be automatically inherited into new simulations from the checkpoint file? 

If so, is _strain_xx_ inheritable? Because I have chosen in _Variables_ to let the displacement inherit the previous simulation, so the strain should be recalculated based on the displacement field at the beginning of the second simulation?  
I'm a bit confused about this, before the _Tensor Mechanics Master Action_ setting came out, I thought my strategy would be to create _AuxVaribales_ and _AuxKernels_ for each stress component and use `initial_from_file_var`, but not do this for _strain_xx_ since the displacement field is already inherited  .  Now after using `Master/all` I am a little confused about this. 
I hope you can clear it up for me, thank you very much!

Best regards,
Daijun

**Reply by GiudGiud** [reply]
these auxiliary variables are only there for visualization purposes. In the equations, the stress/strain material properties are used

**Reply by GiudGiud** [reply]
> so the strain should be recalculated based on the displacement field at the beginning of the second simulation?

it probably should but it's not that simple because there are often history effects (plasticity!) in solid mechanics. So the material properties are not restarted that way. They are restarted properly only using Checkpoint

**Reply by DaijunHU** [reply]
Thank you for your answer!  
If displacements and strain fields are inherited separately, that is, the strain field is inherited as a material property, and the displacement field is inherited manually in  _Variables_ and cannot reflect the historical effect, are there any incompatible parts among them?  
The displacement field cannot reflect whether there is plasticity in it, but it should still be able to make the material enter the plastic segment according to the material constitutive model we set, becaue the calculation is often driven by the displacements. Is this  a kind of duplication or conflict with the inheritance of the strain field?

**Reply by GiudGiud** [reply]
if you use checkpoint to inherit both, displacements and strain will be consistent. 
If you use checkpoint for strain and set the displacements differently, using ICs or another file (with the SolutionUO for example), then they wont be consistent and the results will be affected.
